aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,356 @@
+  static const list_type::const_iterator& unwrap(const ITensorListIterator& it);
+  static ITensorListConstRef iterator_get(const list_type::const_iterator& it);
+};
+} // namespace detail
+
+/*
+ * Wrapper around both boxed and unboxed iterators.
+ *
+ * Currently, a `std::bidirectional_iterator` that wraps those
+ * defined for each of the `ITensorListTag`.
+ *
+ * One should be able to use it, as if it were the unwrapped
+ * iterators themselves.
+ */
+class ITensorListIterator
+    : public std::iterator<std::bidirectional_iterator_tag, at::Tensor> {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...) friend class TORCH_ITENSORLIST_IMPL(TAG);
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_FRIEND_CLASS);
+#undef DEFINE_FRIEND_CLASS
+
+  using unboxed_iterator_type =
+      TORCH_ITENSORLIST_IMPL(Unboxed)::list_type::const_iterator;
+  using boxed_iterator_type =
+      TORCH_ITENSORLIST_IMPL(Boxed)::list_type::const_iterator;
+
+  union Payload {
+    boxed_iterator_type boxed_iterator;
+    unboxed_iterator_type unboxed_iterator;
+    void* _init_ptr;
+    Payload() : _init_ptr(nullptr) {}
+    ~Payload() = default;
+  };
+
+ public:
+  ITensorListIterator() : tag_(ITensorListTag::None) {}
+
+  ITensorListIterator(boxed_iterator_type boxed) : tag_(ITensorListTag::Boxed) {
+    payload_.boxed_iterator = boxed;
+  }
+
+  ITensorListIterator(unboxed_iterator_type unboxed)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed_iterator = unboxed;
+  }
+
+  detail::ITensorListConstRef operator*() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return ImplT::iterator_get(this_); });
+  }
+
+  ITensorListIterator& operator++() {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { ++this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator++(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { ++this_; });
+    return old;
+  }
+
+  ITensorListIterator& operator--() {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator--(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return old;
+  }
+
+  bool operator==(const ITensorListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ITENSORLIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const ITensorListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  Payload payload_;
+  ITensorListTag tag_;
+};
+
+/*
+ * [Note: ITensorList]
+ * Wrapper around boxed and unboxed API containers.
+ *
+ * Tagged union of both API containers:
+ *     - `TensorList`, a.k.a. `ArrayRef<Tensor>` (the unboxed API container)
+ *     - `List<Tensor>` (the boxed API container)
+ *
+ * This container wraps around these two, without incurring in extra overhead
+ * for converting from one to another.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ */
+class ITensorList {
```

Comment:
Blah, I know I was the one who suggested this type name, but I'm having second thoughts, because reading through this reminds me that it is non-owning and it's not obvious at all that ITensorList is non-owning from the name (maybe the I gives it away).

Reply:
ITensorListRef?

Reply:
You have a point. I also would rather have it being obvious in the name. `ITensorListRef` (for this PR) and `IListRef` (for the following PRs) sound good to me.
We still have `TensorList` (which is also non-owning), though.

Reply:
We probably should get someone to change TensorList to TensorArrayRef (it's an ArrayRef<Tensor>, right?)

Reply:
Yes. That's better, I think. So, for consistency reasons, what about `TensorIListRef` and `IListRef`?

Reply:
I think I still prefer having the I up front. Not really sure IListRef is a coherent concept anyway.